goose.go

```diff
@@ -1051,136 +1064,142 @@ func (c *cursor) Remainder() []ast.Stmt {
-	if !ok {
-		ctx.nope(s.Body, "if statement body ends with an assignment")
-		return coq.Binding{}
-	}
+	condExpr := ctx.expr(s.Cond)
 	ife := coq.IfExpr{
-		Cond: ctx.expr(s.Cond),
-		Then: thenExpr,
-	}
-
-	// supported use cases
-	// (1) early return, no else branch; remainder of function is lifted to else
-	// (2) both branches and no remainder
-	//
-	// If the else branch also doesn't return it's not a problem to handle subsequent code,
-	// but that's annoying and we can leave it for later. Maybe the special case
-	// of Else == nil should be supported, though.
-
-	remaining := c.HasNext()
-	bodyEndsWithReturn := endsWithReturn(s.Body)
-	if bodyEndsWithReturn && remaining {
-		if s.Else != nil {
-			ctx.futureWork(s.Else, "else with early return")
-			return coq.Binding{}
-		}
-		ife.Else = ctx.stmts(c.Remainder(), loopVar)
-		return coq.NewAnon(ife)
-	}
-	if !bodyEndsWithReturn && remaining && s.Else == nil {
-		// conditional statement in the middle of a block
-		retUnit := coq.ReturnExpr{coq.Tt}
-		ife.Then = coq.BlockExpr{[]coq.Binding{
-			coq.NewAnon(ife.Then),
-			coq.NewAnon(retUnit),
-		}}
-		ife.Else = retUnit
-		return coq.NewAnon(ife)
-	}
-	elseEndsWithReturn := endsWithReturn(s.Else)
-	// the if expression is last in the surrounding block,
-	// so returns in it become returns from the overall function
-	// OR
-	// neither branch terminates the outer function,
-	// the if expression is just a conditional in the middle
-	if !remaining || (!bodyEndsWithReturn && !elseEndsWithReturn) {
-		if s.Else == nil {
-			ife.Else = coq.ReturnExpr{coq.Tt}
-			return coq.NewAnon(ife)
-		}
-		elseExpr, ok := ctx.stmt(s.Else, c, loopVar).Unwrap()
-		if !ok {
-			ctx.nope(s.Else, "if statement else ends with an assignment")
-			return coq.Binding{}
+		Cond: condExpr,
+	}
+
+	// Determine if we want to propagate our usage (i.e. the availability of control effects)
+	// to the conditional or not. This is not relevant for correctness (stmt generation will
+	// appropriately error when a control effect arises where it is not supported), but Crucial
+	// for supporting a reasonable subset of Go.
+	bodyEndsWithReturn := ctx.endsWithReturn(s.Body)
+
+	// If the "then" branch want to early-return, translate it accordingly, propagating the usage.
+	if bodyEndsWithReturn {
+		// stmts takes care of finalization for us
+		ife.Then = ctx.blockStmt(s.Body, usage)
+		if len(remainder) > 0 {
+			// If there are remaining expressions after this, put them into "else".
+			// Actual "else" must be empty.
+			if s.Else != nil {
+				ctx.futureWork(s.Else, "early return in if with an else branch")
+				return coq.Binding{}
+			}
+			ife.Else = ctx.stmts(remainder, usage)
+		} else {
+			// Remainder is empty; translate "else" branch with our usage.
+			if s.Else == nil {
+				// Be careful to perform proper finalization according to our usage.
+				ife.Else = ctx.stmts([]ast.Stmt{}, usage)
+			} else {
+				ife.Else = ctx.stmts(s.Else.(*ast.BlockStmt).List, usage)
```

Comment:
The old code did some weird things with `Unwrap` here, complaining with "if statement else ends with an assignment". I have no idea what that is about, but this seems to work.